### PR TITLE
Aws upi

### PR DESCRIPTION
**Pull request by yunjiang29**
Support UPI/UPI Private/UPI Proxy/UPI Disconnected&Private cluster installations on AWS/AWS GovCloud.

### CODE REVIEW

File: ci-operator/step-registry/aws/provision/vpc/shared/aws-provision-vpc-shared-commands.sh [hunk 1]
```diff
@@ -13,7 +13,10 @@ TAGS="Key=expirationDate,Value=${EXPIRATION_DATE}"
 
 REGION="${LEASED_RESOURCE}"
 
-curl -L https://raw.githubusercontent.com/openshift/installer/master/upi/aws/cloudformation/01_vpc.yaml -o /tmp/01_vpc.yaml
+vpc_tpl="/tmp/vpc_tpl.yaml"
+
+curl -L https://raw.githubusercontent.com/yunjiang29/ocp-test-data/main/upi_on_aws-cloudformation-templates/01_vpc.yaml -o "${vpc_tpl}" &
+wait "$!"
```
Comment (jianlinliu) [inline ci-operator/step-registry/aws/provision/vpc/shared/aws-provision-vpc-shared-commands.sh:19]: The curl command need `wait` ?

File: ci-operator/step-registry/ipi/conf/aws/sharednetwork/ipi-conf-aws-sharednetwork-commands.sh [hunk 2]
```diff
@@ -32,7 +32,7 @@ fi
 STACK_NAME="${CLUSTER_NAME}-shared-vpc"
 aws --region "${REGION}" cloudformation create-stack \
   --stack-name "${STACK_NAME}" \
-  --template-body "$(cat /tmp/01_vpc.yaml)" \
+  --template-body "$(cat "/tmp/01_vpc.yaml")" \
```
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/conf/aws/sharednetwork/ipi-conf-aws-sharednetwork-commands.sh:35]: need the `"` here?

File: ci-operator/step-registry/upi/install/aws/cluster2/upi-install-aws-cluster2-commands.sh [hunk 3]
```diff
@@ -0,0 +1,603 @@
+#!/bin/bash
```
Comment (jianlinliu) [inline ci-operator/step-registry/upi/install/aws/cluster2/upi-install-aws-cluster2-commands.sh:1]: The reason of introducing this new file is for not touching the existing one - `upi-install-aws-cluster`, right?

File: ci-operator/step-registry/upi/install/aws/cluster2/upi-install-aws-cluster2-commands.sh [hunk 4]
```diff
@@ -0,0 +1,603 @@
...
+
+
+#Actaully if bootkube is still running, all nodes's csr would be signed automatically, but once bootkube is done, would not approve any csr
+#Check `approve-csr` service on bootstrap - /usr/local/bin/approve-csr.sh
+#So have to run some wait and approve here
+wait_and_approve "master" "3" || exit 2
+wait_and_approve "worker" "3" || exit 2
+
+
+cloudfront_base_url=$(head -n 1 /var/run/vault/image-registry/cloudfront-base-url)
```
Comment (jianlinliu) [inline ci-operator/step-registry/upi/install/aws/cluster2/upi-install-aws-cluster2-commands.sh:518]: Is it possible to split the cloudfront configuration part into a separate ref step

File: ci-operator/step-registry/aws/provision/vpc/shared/aws-provision-vpc-shared-commands.sh (hunk 1 above)
Comment (yunjiang29) [reply ci-operator/step-registry/aws/provision/vpc/shared/aws-provision-vpc-shared-commands.sh:19]: I thought that the `curl` command does not require `wait`,  but I got a following error in [this test](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/29050/rehearse-29050-periodic-ci-openshift-verification-tests-master-installer-rehearse-installer-rehearse-aws-qe-upi-disconnected-private/1532232783676575744), the output `/tmp/01_vpc.yaml` is missing:
```
INFO[[20](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/29050/rehearse-29050-periodic-ci-openshift-verification-tests-master-installer-rehearse-installer-rehearse-aws-qe-upi-disconnected-private/1532232783676575744#1:build-log.txt%3A20)22-06-02T05:31:53Z] Logs for container test in pod installer-rehearse-aws-qe-upi-disconnected-private-aws-provision-vpc-disconnected: 
INFO[2022-06-02T05:31:53Z]   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0

100 6113k  100 6113k    0     0  18.8M      0 --:--:-- --:--:-- --:--:-- 18.8M
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  9498  100  9498    0     0  42899      0 --:--:-- --:--:-- --:--:-- 42783
Error parsing parameter '--template-body': Unable to load paramfile file:///tmp/01_vpc.yaml: [Errno 2] No such file or directory: '/tmp/01_vpc.yaml'
```
After adding `wait`, it always get passed.

File: ci-operator/step-registry/ipi/conf/aws/sharednetwork/ipi-conf-aws-sharednetwork-commands.sh (hunk 2 above)
Comment (yunjiang29) [reply ci-operator/step-registry/ipi/conf/aws/sharednetwork/ipi-conf-aws-sharednetwork-commands.sh:35]: I remember that adding `"` is because I met a template format issue https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/29050/rehearse-29050-periodic-ci-openshift-verification-tests-master-installer-rehearse-installer-rehearse-aws-qe/1532215660975755264

File: ci-operator/step-registry/upi/install/aws/cluster2/upi-install-aws-cluster2-commands.sh (hunk 3 above)
Comment (yunjiang29) [reply ci-operator/step-registry/upi/install/aws/cluster2/upi-install-aws-cluster2-commands.sh:1]: yes, both of install and destroy scripts can not be re-used.

File: ci-operator/step-registry/upi/install/aws/cluster2/upi-install-aws-cluster2-commands.sh (hunk 4 above)
Comment (yunjiang29) [reply ci-operator/step-registry/upi/install/aws/cluster2/upi-install-aws-cluster2-commands.sh:518]: Sure, but we still need to config an empty dir for Image Registry https://docs.openshift.com/container-platform/4.10/installing/installing_aws/installing-aws-user-infra.html#installation-registry-storage-non-production_installing-aws-user-infra

File: ci-operator/step-registry/aws/provision/vpc/shared/aws-provision-vpc-shared-commands.sh (hunk 1 above)
Comment (jianlinliu) [reply ci-operator/step-registry/aws/provision/vpc/shared/aws-provision-vpc-shared-commands.sh:19]: okay. Get it.

File: ci-operator/step-registry/ipi/conf/aws/sharednetwork/ipi-conf-aws-sharednetwork-commands.sh (hunk 2 above)
Comment (jianlinliu) [reply ci-operator/step-registry/ipi/conf/aws/sharednetwork/ipi-conf-aws-sharednetwork-commands.sh:35]: hmm, I guess that is some flake. Can we remove it for one more try.

File: ci-operator/step-registry/upi/install/aws/cluster2/upi-install-aws-cluster2-commands.sh (hunk 4 above)
Comment (jianlinliu) [reply ci-operator/step-registry/upi/install/aws/cluster2/upi-install-aws-cluster2-commands.sh:518]: I prefer to keep the install process as simple as possible, allow workflow developer to opt-in image-registry storage configuration as desire, also avoid binding vault with the upi ref step.

File: ci-operator/step-registry/ipi/conf/aws/sharednetwork/ipi-conf-aws-sharednetwork-commands.sh (hunk 2 above)
Comment (yunjiang29) [reply ci-operator/step-registry/ipi/conf/aws/sharednetwork/ipi-conf-aws-sharednetwork-commands.sh:35]: @jianlinliu  this must be a mistake, this file should not be changed, the ref is not used in workflows. Reverted.
Thanks

File: ci-operator/step-registry/upi/install/aws/cluster2/upi-install-aws-cluster2-commands.sh (hunk 4 above)
Comment (yunjiang29) [reply ci-operator/step-registry/upi/install/aws/cluster2/upi-install-aws-cluster2-commands.sh]: updated, patched an empty dir.

File: ci-operator/step-registry/aws/upi-templates-generate/OWNERS [hunk 5]
```diff
@@ -0,0 +1,5 @@
+approvers:
+- patrickdillon
+- wking
```
Comment (wking) [inline ci-operator/step-registry/aws/upi-templates-generate/OWNERS:3]: Can you drop me?  I don't think I have much to contribute to UPI templates anymore, certainly not at a "likely to be able to provide timely review/direction" level.

File: ci-operator/step-registry/aws/upi-templates-generate/aws-upi-templates-generate-commands.sh [hunk 6]
```diff
@@ -0,0 +1,2780 @@
...
+# 05_cluster_master_nodes.yaml
+# 06_cluster_worker_node.yaml
+# 97_apps_ingress-elb_dns.yaml
+# 99_int_service_instance.yaml
+
+#
+# template: 01_vpc.yaml
+# 
+echo "Creating 01_vpc.yaml"
+cat > "${SHARED_DIR}/01_vpc.yaml" <<EOF
```
Comment (wking) [inline ci-operator/step-registry/aws/upi-templates-generate/aws-upi-templates-generate-commands.sh:17]: These feel a lot like [the installer UPI templates][1].  If they're different, it would be nice to have the step docs explain how, and also why we decided to put the different manifests in a step instead of alongside the existing installer templates.  If they're the same, it's not clear what we gain by duplicating them in a new location.

[1]: https://github.com/openshift/installer/tree/master/upi/aws/cloudformation

File: ci-operator/step-registry/aws/upi-templates-generate/aws-upi-templates-generate-commands.sh [hunk 7]
```diff
@@ -0,0 +1,2780 @@
...
+  InternalServiceTargetGroupArn:
+    Description: ARN of internal service target group.
+    Value: !Ref InternalServiceTargetGroup
+EOF
+
+#
+# template: 03_cluster_security.yaml
+# 
+echo "Creating 03_cluster_security.yaml"
+cat > "${SHARED_DIR}/03_cluster_security.yaml" <<EOF
```
Comment (wking) [inline ci-operator/step-registry/aws/upi-templates-generate/aws-upi-templates-generate-commands.sh:1153]: The only consumer for this one appears to be `upi-install-aws-cluster-custom`.  I'm unclear on the benefit to declaring the template here, storing it in the space-limited `SHARED_DIR`, and then only consuming it in a single other step is.  Wouldn't it be easier and more efficient to declare it locally in the step that needs it?

If the goal is just reducing clutter in the consuming step, it may be worth reviving [DPTP-1259][1] (internal ticket; sorry external folks).

[1]: https://issues.redhat.com/browse/DPTP-1259?

File: ci-operator/step-registry/upi/deprovision/aws/stage1/upi-deprovision-aws-stage1-commands.sh [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    --query 'StackResources[?ResourceType==`AWS::EC2::Instance`].PhysicalResourceId' --output text | sed 's,\t,\n,g' > /tmp/node-provider-IDs
+for INDEX in 0 1 2
+do
+    aws --region $AWS_REGION cloudformation describe-stack-resources --stack-name "${CLUSTER_NAME}-compute-${INDEX}" \
+    --query 'StackResources[].PhysicalResourceId' --output text | cut -d, -f1 >> /tmp/node-provider-IDs
+done
+
+while IFS= read -r i; do
+    mkdir -p "${ARTIFACT_DIR}/nodes/${i}"
+    queue ${ARTIFACT_DIR}/nodes/$i/console aws ec2 get-console-output --instance-id "${i}"
```
Comment (wking) [inline ci-operator/step-registry/upi/deprovision/aws/stage1/upi-deprovision-aws-stage1-commands.sh:45]: The `documentation` for this step is talking about deprovisioning, and this seems orthogonal.  There's also an existing `gather-aws-console` step that seems like it will do what you're looking for here, so perhaps you can drop this portion and plug `ref: gather-aws-console` into your chains instead?

File: ci-operator/step-registry/upi/deprovision/aws/stage1/upi-deprovision-aws-stage1-commands.sh [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+export AWS_REGION
+export AWS_DEFAULT_REGION
+CLUSTER_NAME=$(cat ${SHARED_DIR}/CLUSTER_NAME)
+
+aws --region $AWS_REGION cloudformation describe-stack-resources --stack-name "${CLUSTER_NAME}-control-plane" \
+    --query 'StackResources[?ResourceType==`AWS::EC2::Instance`].PhysicalResourceId' --output text | sed 's,\t,\n,g' > /tmp/node-provider-IDs
+for INDEX in 0 1 2
+do
+    aws --region $AWS_REGION cloudformation describe-stack-resources --stack-name "${CLUSTER_NAME}-compute-${INDEX}" \
+    --query 'StackResources[].PhysicalResourceId' --output text | cut -d, -f1 >> /tmp/node-provider-IDs
```
Comment (wking) [inline ci-operator/step-registry/upi/deprovision/aws/stage1/upi-deprovision-aws-stage1-commands.sh:40]: The current approach for `gather-aws-console` is to populate `${SHARED_DIR}/aws-instance-ids.txt` as you create the instances, e.g. [here][1].  That way, `gather-aws-console` can focus on going from instance IDs to logs, and doesn't have to understand the various ways previous steps may have provisioned instances (Terraform, CloudFormation, etc.).

[1]: https://github.com/openshift/release/blob/7717497fc639610d4b537752d50408236f15a610/ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh#L19

File: ci-operator/step-registry/upi/deprovision/aws/stage1/upi-deprovision-aws-stage1-ref.yaml [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+  from: upi-installer
+  grace_period: 10m
+  commands: upi-deprovision-aws-stage1-commands.sh
+  resources:
+    requests:
+      cpu: 10m
+      memory: 100Mi
+  documentation: >-
+    This step delets stacks before running `openshift-install destroy` command
+    e.g. *-security *-bootstrap, *-control-plane *-compute* *apps-dns stacks.
```
Comment (wking) [inline ci-operator/step-registry/upi/deprovision/aws/stage1/upi-deprovision-aws-stage1-ref.yaml:12]: instead of being opinionated about what you're likely to pick up, it might be easier to maintain if you doc the `${SHARED_DIR}/to_be_removed_cf_list_stage_1` input format, and then leave it to folks wiring up workflows to make the match between "step A creates `${SHARED_DIR}/${pathA}`, and step B consumes it".

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/upi/deprovision/cucushift-installer-rehearse-aws-upi-deprovision-chain.yaml [hunk 11]
```diff
@@ -0,0 +1,11 @@
+chain:
+  as: cucushift-installer-rehearse-aws-upi-deprovision
+  steps:
+  - chain: gather
+  - ref: upi-deprovision-aws-stage1
+  - ref: ipi-deprovision-deprovision
+  - ref: upi-deprovision-aws-stage2
```
Comment (wking) [inline ci-operator/step-registry/cucushift/installer/rehearse/aws/upi/deprovision/cucushift-installer-rehearse-aws-upi-deprovision-chain.yaml:7]: I don't understand the need to split `stage1` and `stage2` with the IPI deprovisioner in the middle.  Our [uninstall docs][1] have no section for clusters built on user-provisioned infrastructure.  It seems like I commited a `delete-stack` , `destroy cluster`, `delete-stack`-again pattern in 602791d72c475d5878676305eca1b27274cffdef, although I neglected to record my motivation in the commit message.  Maybe it was:

1. Handwave, maybe stack deletion just works.  Try it.
2. Maybe it didn't work, because some cluster-created resource is blocking things?  Try a `destroy cluster`, in case there is anything cluster-created (or tagged as cluster-owned in a stack) that the installer can remove.
3. Hopefully that got any blockers out of the way.  Try the stacks again.

Depending on how solidly user-provisioned the workflow is intending to be, 2 and 3 may be unecessary complication.

[1]: https://docs.openshift.com/container-platform/4.11/installing/installing_aws/uninstalling-cluster-aws.html

File: ci-operator/step-registry/upi/install/aws/cluster-custom/upi-install-aws-cluster-custom-ref.yaml [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+  - name: ADD_INGRESS_RECORDS_MANUALLY
+    default: "no"
+  dependencies:
+  - name: "release:latest"
+    env: OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE
+  documentation: >-
+    This step deploys a UPI cluster to the CI AWS project.
+    It allows you to create app dns records manually via setting ADD_INGRESS_RECORDS_MANUALLY
+    And supoprts private cluster and disconnected & private cluster.
+    The VPC and basion host are created by aws-provision-vpc* and aws-provision-bastionhost
```
Comment (wking) [inline ci-operator/step-registry/upi/install/aws/cluster-custom/upi-install-aws-cluster-custom-ref.yaml:26]: there's a lot of overlap, and also a lot of divergence, between this new step and the existing `upi-install-aws-cluster`.  I'm not a maintainer of either, so :shrug: y'all can do whatever makes most sense to you, but it seems like the two maintainer sets may end up doing duplicate work, and anyone looking for an AWS UPI install step to consume will have a hard time selecting which is the best fit for their goals.

File: ci-operator/step-registry/aws/upi-templates-generate/aws-upi-templates-generate-commands.sh (hunk 7 above)
Comment (yunjiang29) [reply ci-operator/step-registry/aws/upi-templates-generate/aws-upi-templates-generate-commands.sh:1153]: > If the goal is just reducing clutter in the consuming step, it may be worth reviving [DPTP-1259](https://issues.redhat.com/browse/DPTP-1259?) (internal ticket; sorry external folks).

Yes, this is the goal of splitting out the template creation step.
Most of the templates are used in UPI install step only, but it looks like there is still no agreement on whether to implement [DPTP-1259](https://issues.redhat.com/browse/DPTP-1259), I will create templates locally in this PR.

File: ci-operator/step-registry/aws/upi-templates-generate/OWNERS (hunk 5 above)
Comment (yunjiang29) [reply ci-operator/step-registry/aws/upi-templates-generate/OWNERS:3]: OK, will do.

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/upi/deprovision/cucushift-installer-rehearse-aws-upi-deprovision-chain.yaml (hunk 11 above)
Comment (yunjiang29) [reply ci-operator/step-registry/cucushift/installer/rehearse/aws/upi/deprovision/cucushift-installer-rehearse-aws-upi-deprovision-chain.yaml:7]: Thank you @wking for providing so detailed info, I will try and make the deprovision steps as easy as possible.

File: ci-operator/step-registry/upi/install/aws/cluster-custom/upi-install-aws-cluster-custom-ref.yaml (hunk 12 above)
Comment (yunjiang29) [reply ci-operator/step-registry/upi/install/aws/cluster-custom/upi-install-aws-cluster-custom-ref.yaml:26]: Yes, it's true that there are a lot of overlaps and differences.
For our tests, we need to cover many test configurations, so we did some customization, but we don't want to break up the existing processes.

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/upi/deprovision/cucushift-installer-rehearse-aws-upi-deprovision-chain.yaml (hunk 11 above)
Comment (yunjiang29) [reply ci-operator/step-registry/cucushift/installer/rehearse/aws/upi/deprovision/cucushift-installer-rehearse-aws-upi-deprovision-chain.yaml:7]: @wking The `*.apps.[cluster-name].[base-domainname]` record created by ingress will cause deleting stack `02_cluster_infra.yaml` failed, that's why I put it in  `stage2`,  after the `destroy` command. 

Other resources, like s3 buckets created image registry, IAM policies created by CR, also need to be deleted by `openshift-install destroy`. 

I think that's why you used `delete-stack` , `destroy cluster`, `delete-stack`-again pattern in https://github.com/openshift/release/commit/602791d72c475d5878676305eca1b27274cffdef, also the similar process is being used in [existing UPI post action](https://github.com/openshift/release/blob/master/ci-operator/step-registry/upi/aws/post/upi-aws-post-chain.yaml#L5-L7).

To make the process easy to understand, I will use the similar one:
1. delete all stacks (deleting infrastructure stack will fail)
2. `openshift-install destroy` 
3. delete failed (or all)  stacks again.

File: ci-operator/step-registry/aws/upi-templates-generate/aws-upi-templates-generate-commands.sh (hunk 7 above)
Comment (yunjiang29) [reply ci-operator/step-registry/aws/upi-templates-generate/aws-upi-templates-generate-commands.sh]: All templates are generated in step `upi-install-aws-cluster-custom` locally now.

File: ci-operator/step-registry/aws/upi-templates-generate/aws-upi-templates-generate-commands.sh (hunk 6 above)
Comment (yunjiang29) [reply ci-operator/step-registry/aws/upi-templates-generate/aws-upi-templates-generate-commands.sh]: Updated. All templates are generated in step `upi-install-aws-cluster-custom` locally now and removed VPC parts.

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/upi/deprovision/cucushift-installer-rehearse-aws-upi-deprovision-chain.yaml (hunk 11 above)
Comment (yunjiang29) [reply ci-operator/step-registry/cucushift/installer/rehearse/aws/upi/deprovision/cucushift-installer-rehearse-aws-upi-deprovision-chain.yaml]: I reused existing `upi-deprovision-aws` and `upi-deprovision-aws-check-destroy` steps, and removed `stage[12]`.

Comment (wking) [reply ci-operator/step-registry/cucushift/installer/rehearse/aws/upi/deprovision/cucushift-installer-rehearse-aws-upi-deprovision-chain.yaml]: ok, pinning this down so I don't forget:
1. First stack deletion.  It won't complete (blocked by cluster-created resources like ingress DNS records, etc.).  But we need to terminate instances and otherwise hobble the cluster as much as we can, so that `openshift-install destroy` will be able to delete cluster-created resources.  If we opened with `openshift-install destroy`, the cluster might remain healthy enough to create replacement resources.
2. `openshift-install destroy` to remove cluster-created resources, which the stacks don't know about.
3. Remove the stacks, now that they're unblocked.

File: ci-operator/step-registry/upi/install/aws/cluster-custom/upi-install-aws-cluster-custom-ref.yaml (hunk 12 above)
Comment (wking) [reply ci-operator/step-registry/upi/install/aws/cluster-custom/upi-install-aws-cluster-custom-ref.yaml:26]: so docs like "use this step if you need more flexibility than upi-install-aws-cluster provides"?

**Comment by yunjiang29** [comment]
@wking PTAL, let me know if any further comments,  thanks for your review!
